File: packages/protocol/src/frames.ts
import { z } from 'zod';
import type { Json } from './json.js';
import { PROTOCOL_VERSION } from './version.js';

const jsonSchema: z.ZodType<Json> = z.lazy(() =>
  z.union([
    z.string(),
    z.number(),
    z.boolean(),
    z.null(),
    z.array(jsonSchema),
    z.record(jsonSchema)
  ])
);

const v = z.literal(PROTOCOL_VERSION);

export const connectionStateSchema = z.enum([
  'connecting',
  'connected',
  'reconnecting',
  'unreachable',
  'left'
]);

export const participantStateSchema = z.object({
  muted: z.boolean(),
  onHold: z.boolean()
});

export const partialStateSchema = z.object({
  muted: z.boolean().optional(),
  onHold: z.boolean().optional()
});

export const participantSchema = z.object({
  participantId: z.string(),
  joinedAt: z.number(),
  metadata: z.record(jsonSchema),
  connectionState: connectionStateSchema,
  state: participantStateSchema
});

export const iceCandidateSchema = z.object({
  candidate: z.string(),
  sdpMid: z.string().nullable(),
  sdpMLineIndex: z.number().nullable()
});

export const iceServerSchema = z
  .object({
    urls: z.union([z.string(), z.array(z.string())]),
    username: z.string().optional(),
    credential: z.string().optional(),
    realm: z.string().optional()
  })
  .passthrough();

export const joinFrameSchema = z
  .object({
    v,
    type: z.literal('join'),
    token: z.string()
  })
  .strict();

export const leaveFrameSchema = z
  .object({
    v,
    type: z.literal('leave')
  })
  .strict();

export const clientStateFrameSchema = z
  .object({
    v,
    type: z.literal('state'),
    state: partialStateSchema
  })
  .strict();

export const offerSendSchema = z
  .object({
    v,
    type: z.literal('offer'),
    to: z.string(),
    sdp: z.string()
  })
  .strict();

export const answerSendSchema = z
  .object({
    v,
    type: z.literal('answer'),
    to: z.string(),
    sdp: z.string()
  })
  .strict();

export const iceSendSchema = z
  .object({
    v,
    type: z.literal('ice'),
    to: z.string(),
    candidate: iceCandidateSchema
  })
  .strict();

export const offerDeliveredSchema = z
  .object({
    v,
    type: z.literal('offer'),
    from: z.string(),
    sdp: z.string()
  })
  .strict();

export const answerDeliveredSchema = z
  .object({
    v,
    type: z.literal('answer'),
    from: z.string(),
    sdp: z.string()
  })
  .strict();

export const iceDeliveredSchema = z
  .object({
    v,
    type: z.literal('ice'),
    from: z.string(),
    candidate: iceCandidateSchema
  })
  .strict();

export const readyFrameSchema = z.object({
  v,
  type: z.literal('ready'),
  self: participantSchema,
  peers: z.array(participantSchema),
  room: z.object({
    roomId: z.string(),
    startedAt: z.number().nullable(),
    maxParticipants: z.number()
  }),
  iceServers: z.array(iceServerSchema),
  serverNow: z.number(),
  resumed: z.boolean()
});

export const participantJoinedFrameSchema = z.object({
  v,
  type: z.literal('participant-joined'),
  participant: participantSchema
});

export const participantStateFrameSchema = z.object({
  v,
  type: z.literal('participant-state'),
  participantId: z.string(),
  state: partialStateSchema
});

export const participantUnreachableFrameSchema = z.object({
  v,
  type: z.literal('participant-unreachable'),
  participantId: z.string()
});

export const participantLeftFrameSchema = z.object({
  v,
  type: z.literal('participant-left'),
  participantId: z.string()
});

export const roomStartedFrameSchema = z.object({
  v,
  type: z.literal('room-started'),
  startedAt: z.number()
});

export const errorFrameSchema = z.object({
  v,
  type: z.literal('error'),
  code: z.string(),
  message: z.string()
});

export const clientFrameSchema = z.discriminatedUnion('type', [
  joinFrameSchema,
  leaveFrameSchema,
  clientStateFrameSchema,
  offerSendSchema,
  answerSendSchema,
  iceSendSchema
]);

export const serverFrameSchema = z.discriminatedUnion('type', [
  readyFrameSchema,
  participantJoinedFrameSchema,
  participantStateFrameSchema,
  participantUnreachableFrameSchema,
  participantLeftFrameSchema,
  roomStartedFrameSchema,
  errorFrameSchema,
  offerDeliveredSchema,
  answerDeliveredSchema,
  iceDeliveredSchema
]);

export type JoinFrame = z.infer<typeof joinFrameSchema>;
export type LeaveFrame = z.infer<typeof leaveFrameSchema>;
export type ClientStateFrame = z.infer<typeof clientStateFrameSchema>;
export type OfferSendFrame = z.infer<typeof offerSendSchema>;
export type AnswerSendFrame = z.infer<typeof answerSendSchema>;
export type IceSendFrame = z.infer<typeof iceSendSchema>;
export type OfferDeliveredFrame = z.infer<typeof offerDeliveredSchema>;
export type AnswerDeliveredFrame = z.infer<typeof answerDeliveredSchema>;
export type IceDeliveredFrame = z.infer<typeof iceDeliveredSchema>;
export type ReadyFrame = z.infer<typeof readyFrameSchema>;
export type ParticipantJoinedFrame = z.infer<typeof participantJoinedFrameSchema>;
export type ParticipantStateFrame = z.infer<typeof participantStateFrameSchema>;
export type ParticipantUnreachableFrame = z.infer<
  typeof participantUnreachableFrameSchema
>;
export type ParticipantLeftFrame = z.infer<typeof participantLeftFrameSchema>;
export type RoomStartedFrame = z.infer<typeof roomStartedFrameSchema>;
export type ErrorFrame = z.infer<typeof errorFrameSchema>;
export type ClientFrame = z.infer<typeof clientFrameSchema>;
export type ServerFrame = z.infer<typeof serverFrameSchema>;

export function isVersionMismatch(raw: unknown): boolean {
  return (
    typeof raw === 'object' &&
    raw !== null &&
    'v' in raw &&
    (raw as { v: unknown }).v !== PROTOCOL_VERSION
  );
}

export function parseClientFrame(raw: unknown) {
  return clientFrameSchema.safeParse(raw);
}

export function parseServerFrame(raw: unknown) {
  return serverFrameSchema.safeParse(raw);
}

File: android/engine/bin/main/io/sauti/engine/Frames.kt
package io.sauti.engine

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

const val PROTOCOL_VERSION = 1

sealed interface ServerFrame {
    data class Ready(
        val self: Participant,
        val peers: List<Participant>,
        val room: RoomInfo,
        val iceServers: List<IceServer>,
        val serverNow: Long,
        val resumed: Boolean
    ) : ServerFrame

    data class ParticipantJoined(val participant: Participant) : ServerFrame
    data class ParticipantStateChanged(val participantId: String, val state: PartialState) : ServerFrame
    data class ParticipantUnreachable(val participantId: String) : ServerFrame
    data class ParticipantLeft(val participantId: String) : ServerFrame
    data class RoomStarted(val startedAt: Long) : ServerFrame
    data class ErrorFrame(val code: String, val message: String) : ServerFrame
    data class Offer(val from: String, val sdp: String) : ServerFrame
    data class Answer(val from: String, val sdp: String) : ServerFrame
    data class Ice(val from: String, val candidate: IceCandidate) : ServerFrame
}

sealed interface FrameParseResult {
    data class Parsed(val frame: ServerFrame) : FrameParseResult
    data class VersionMismatch(val version: Int?) : FrameParseResult
    object Invalid : FrameParseResult
}

object FrameCodec {
    private val json = Json { ignoreUnknownKeys = true }

    fun encodeJoin(token: String): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "join")
        put("token", token)
    }.toString()

    fun encodeLeave(): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "leave")
    }.toString()

    fun encodeState(muted: Boolean?, onHold: Boolean?): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "state")
        put("state", buildJsonObject {
            if (muted != null) put("muted", muted)
            if (onHold != null) put("onHold", onHold)
        })
    }.toString()

    fun encodeOffer(to: String, sdp: String): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "offer")
        put("to", to)
        put("sdp", sdp)
    }.toString()

    fun encodeAnswer(to: String, sdp: String): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "answer")
        put("to", to)
        put("sdp", sdp)
    }.toString()

    fun encodeIce(to: String, candidate: IceCandidate): String = buildJsonObject {
        put("v", PROTOCOL_VERSION)
        put("type", "ice")
        put("to", to)
        put("candidate", encodeCandidate(candidate))
    }.toString()

    private fun encodeCandidate(candidate: IceCandidate): JsonObject = buildJsonObject {
        put("candidate", candidate.candidate)
        if (candidate.sdpMid != null) put("sdpMid", candidate.sdpMid) else put("sdpMid", JsonNull)
        if (candidate.sdpMLineIndex != null) put("sdpMLineIndex", candidate.sdpMLineIndex) else put("sdpMLineIndex", JsonNull)
    }

    fun parse(text: String): FrameParseResult {
        val root = try {
            json.parseToJsonElement(text).jsonObject
        } catch (error: Exception) {
            return FrameParseResult.Invalid
        }

        val version = root["v"]?.jsonPrimitive?.content?.toIntOrNull()
        if (version != PROTOCOL_VERSION) {
            return FrameParseResult.VersionMismatch(version)
        }

        val type = root["type"]?.jsonPrimitive?.content ?: return FrameParseResult.Invalid

        val frame = when (type) {
            "ready" -> parseReady(root)
            "participant-joined" -> ServerFrame.ParticipantJoined(parseParticipant(root.getValue("participant").jsonObject))
            "participant-state" -> ServerFrame.ParticipantStateChanged(
                root.getValue("participantId").jsonPrimitive.content,
                parsePartialState(root.getValue("state").jsonObject)
            )
            "participant-unreachable" -> ServerFrame.ParticipantUnreachable(root.getValue("participantId").jsonPrimitive.content)
            "participant-left" -> ServerFrame.ParticipantLeft(root.getValue("participantId").jsonPrimitive.content)
            "room-started" -> ServerFrame.RoomStarted(root.getValue("startedAt").jsonPrimitive.content.toLong())
            "error" -> ServerFrame.ErrorFrame(
                root.getValue("code").jsonPrimitive.content,
                root.getValue("message").jsonPrimitive.content
            )
            "offer" -> ServerFrame.Offer(root.getValue("from").jsonPrimitive.content, root.getValue("sdp").jsonPrimitive.content)
            "answer" -> ServerFrame.Answer(root.getValue("from").jsonPrimitive.content, root.getValue("sdp").jsonPrimitive.content)
            "ice" -> ServerFrame.Ice(
                root.getValue("from").jsonPrimitive.content,
                parseCandidate(root.getValue("candidate").jsonObject)
            )
            else -> return FrameParseResult.Invalid
        }
        return FrameParseResult.Parsed(frame)
    }

    private fun parseReady(root: JsonObject): ServerFrame.Ready {
        val roomObj = root.getValue("room").jsonObject
        val startedAt = roomObj["startedAt"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content?.toLong()
        return ServerFrame.Ready(
            self = parseParticipant(root.getValue("self").jsonObject),
            peers = root.getValue("peers").jsonArray.map { parseParticipant(it.jsonObject) },
            room = RoomInfo(
                roomId = roomObj.getValue("roomId").jsonPrimitive.content,
                startedAt = startedAt,
                maxParticipants = roomObj.getValue("maxParticipants").jsonPrimitive.content.toInt()
            ),
            iceServers = (root["iceServers"]?.jsonArray ?: buildJsonArray { }).map { parseIceServer(it.jsonObject) },
            serverNow = root.getValue("serverNow").jsonPrimitive.content.toLong(),
            resumed = root["resumed"]?.jsonPrimitive?.content?.toBoolean() ?: false
        )
    }

    private fun parseParticipant(obj: JsonObject): Participant {
        val stateObj = obj["state"]?.jsonObject
        return Participant(
            participantId = obj.getValue("participantId").jsonPrimitive.content,
            joinedAt = obj["joinedAt"]?.jsonPrimitive?.content?.toLong() ?: 0L,
            metadata = obj["metadata"]?.jsonObject?.let { readMetadata(it) } ?: emptyMap(),
            connectionState = parseConnectionState(obj["connectionState"]?.jsonPrimitive?.content),
            state = ParticipantState(
                muted = stateObj?.get("muted")?.jsonPrimitive?.content?.toBoolean() ?: false,
                onHold = stateObj?.get("onHold")?.jsonPrimitive?.content?.toBoolean() ?: false
            )
        )
    }

    private fun readMetadata(obj: JsonObject): Map<String, JsonElement> = obj.toMap()

    private fun parsePartialState(obj: JsonObject): PartialState = PartialState(
        muted = obj["muted"]?.jsonPrimitive?.content?.toBoolean(),
        onHold = obj["onHold"]?.jsonPrimitive?.content?.toBoolean()
    )

    private fun parseConnectionState(raw: String?): ConnectionState = when (raw) {
        "connecting" -> ConnectionState.CONNECTING
        "connected" -> ConnectionState.CONNECTED
        "reconnecting" -> ConnectionState.RECONNECTING
        "unreachable" -> ConnectionState.UNREACHABLE
        "left" -> ConnectionState.LEFT
        else -> ConnectionState.CONNECTING
    }

    private fun parseIceServer(obj: JsonObject): IceServer {
        val urlsElement = obj["urls"]
        val urls = when {
            urlsElement is JsonPrimitive -> listOf(urlsElement.content)
            urlsElement != null -> urlsElement.jsonArray.map { it.jsonPrimitive.content }
            else -> emptyList()
        }
        return IceServer(
            urls = urls,
            username = obj["username"]?.jsonPrimitive?.content,
            credential = obj["credential"]?.jsonPrimitive?.content
        )
    }

    private fun parseCandidate(obj: JsonObject): IceCandidate = IceCandidate(
        candidate = obj.getValue("candidate").jsonPrimitive.content,
        sdpMid = obj["sdpMid"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
        sdpMLineIndex = obj["sdpMLineIndex"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content?.toInt()
    )
}
